File: app/src/main/java/org/ligi/blexplorer/scan/DeviceViewHolder.java
package org.ligi.blexplorer.scan;

import android.app.Activity;
import android.bluetooth.BluetoothDevice;
import android.content.Intent;
import android.os.ParcelUuid;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.SparseArray;
import android.view.View;
import android.widget.TextView;

import butterknife.Bind;
import butterknife.ButterKnife;
import java.math.BigInteger;
import org.ligi.blexplorer.App;
import org.ligi.blexplorer.R;
import org.ligi.blexplorer.services.DeviceServiceExploreActivity;
import org.ligi.blexplorer.util.ManufacturerRecordParserFactory;
import org.ligi.blexplorer.util.from_lollipop.ScanRecord;
import static org.ligi.blexplorer.util.DevicePropertiesDescriber.describeBondState;
import static org.ligi.blexplorer.util.DevicePropertiesDescriber.describeType;

public class DeviceViewHolder extends RecyclerView.ViewHolder {

    @Bind(R.id.address)
    public TextView address;

    @Bind(R.id.name)
    public TextView name;

    @Bind(R.id.bondstate)
    public TextView bondstate;

    @Bind(R.id.type)
    public TextView type;

    @Bind(R.id.last_seen)
    public TextView last_seen;

    @Bind(R.id.rssi)
    public TextView rssi;

    @Bind(R.id.scan_record)
    public TextView scan_record;

    public BluetoothDevice device;

    public void applyDevice(BluetoothDevice newDevice, DeviceListActivity.DeviceExtras extras) {
        device = newDevice;
        name.setText(TextUtils.isEmpty(device.getName()) ? "no name" : device.getName());
        rssi.setText("" + extras.rssi + "db");
        last_seen.setText("" + (System.currentTimeMillis() - extras.last_seen) / 1000 + "s");
        address.setText(device.getAddress());

        final ScanRecord scanRecord = ScanRecord.parseFromBytes(extras.scanRecord);
        String scanRecordStr = "";
        if (scanRecord.getServiceUuids() != null) {
            for (final ParcelUuid parcelUuid : scanRecord.getServiceUuids()) {
                scanRecordStr += parcelUuid.toString() + "\n";
            }
        }

        final SparseArray<byte[]> manufacturerSpecificData = scanRecord.getManufacturerSpecificData();

        for (int i = 0; i < manufacturerSpecificData.size(); i++) {
            final int key = manufacturerSpecificData.keyAt(i);
            ManufacturerRecordParserFactory.ManufacturerParserBase p = ManufacturerRecordParserFactory.parse(key, manufacturerSpecificData.get(key));
            if (p == null) {
                scanRecordStr += key + "=" + new BigInteger(1, manufacturerSpecificData.get(key)).toString(16) + "\n";
            } else {
                scanRecordStr += p.getKeyDescriptor() + " = {\n" + p.toString() + "}\n";
            }
        }

        for (final ParcelUuid parcelUuid : scanRecord.getServiceData().keySet()) {
            scanRecordStr += parcelUuid + "=" + new BigInteger(1, scanRecord.getServiceData().get(parcelUuid)).toString(16) + "\n";
        }

        scan_record.setText(scanRecordStr);

        type.setText(describeType(device));
        bondstate.setText(describeBondState(device));
    }

    public void installOnClickListener(final Activity activity) {
        itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {
                final Intent intent = new Intent(activity, DeviceServiceExploreActivity.class);
                App.device = device;
                activity.startActivity(intent);
            }
        });
    }

    public DeviceViewHolder(final View itemView) {
        super(itemView);

        ButterKnife.bind(this, itemView);
    }
}

File: android/src/main/java/org/ligi/blexplorer/scan/DeviceViewHolder.kt
package org.ligi.blexplorer.scan

import android.app.Activity
import android.bluetooth.BluetoothDevice
import android.content.Intent
import androidx.recyclerview.widget.RecyclerView
import android.text.TextUtils
import org.ligi.blexplorer.App
import org.ligi.blexplorer.databinding.ItemDeviceBinding
import org.ligi.blexplorer.services.DeviceServiceExploreActivity
import org.ligi.blexplorer.util.DevicePropertiesDescriber.describeBondState
import org.ligi.blexplorer.util.DevicePropertiesDescriber.describeType
import org.ligi.blexplorer.util.ManufacturerRecordParserFactory
import org.ligi.blexplorer.util.from_lollipop.ScanRecord
import java.math.BigInteger

class DeviceViewHolder(private val binding: ItemDeviceBinding) : androidx.recyclerview.widget.RecyclerView.ViewHolder(binding.root) {

    lateinit var device: BluetoothDevice

    fun applyDevice(newDevice: BluetoothDevice, extras: DeviceListActivity.DeviceExtras) {
        device = newDevice
        binding.name.text = if (TextUtils.isEmpty(device.name)) "no name" else device.name
        binding.rssi.text = "${extras.rssi}db"
        binding.lastSeen.text = "" + (System.currentTimeMillis() - extras.last_seen) / 1000 + "s"
        binding.address.text = device.address

        val scanRecord = ScanRecord.parseFromBytes(extras.scanRecord)
        var scanRecordStr = ""
        if (scanRecord.serviceUuids != null) {
            for (parcelUuid in scanRecord.serviceUuids) {
                scanRecordStr += parcelUuid.toString() + "\n"
            }
        }

        val manufacturerSpecificData = scanRecord.manufacturerSpecificData

        (0 until manufacturerSpecificData.size())
                .map { manufacturerSpecificData.keyAt(it) }
                .forEach {key ->
                    val p = ManufacturerRecordParserFactory.parse(key, manufacturerSpecificData.get(key), device)
                    if (p == null) {
                        scanRecordStr += "$key=" + BigInteger(1, manufacturerSpecificData.get(key)).toString(16) + "\n"
                    } else {
                        scanRecordStr += p.keyDescriptor + " = {\n" + p.toString() + "}\n"
                        if (!TextUtils.isEmpty(p.getName(device))) {
                            binding.name.text = p.getName(device)
                        }
                    }
                }

        for (parcelUuid in scanRecord.serviceData.keys) {
            scanRecordStr += "$parcelUuid=" + BigInteger(1, scanRecord.serviceData[parcelUuid]).toString(16) + "\n"
        }

        binding.scanRecord.text = scanRecordStr

        binding.type.text = describeType(device)
        binding.bondstate.text = describeBondState(device)
    }

    fun installOnClickListener(activity: Activity) {
        itemView.setOnClickListener {
            val intent = Intent(activity, DeviceServiceExploreActivity::class.java)
            App.device = device
            activity.startActivity(intent)
        }
    }

}
